### PR TITLE
Add Rust Support and Advanced Logger Crate [Rebase & FF]

### LINKED ISSUE #294: [Feature]: Integrate Rust build support with library crate
### Feature Overview

This issue tracks integrating Rust support into the repo and testing the build with a simple library crate. The infrastructure will be reused elsewhere in Project Mu with less effort in the future.

### Solution Overview

- Place reusable content in [mu_devops](https://github.com/microsoft/mu_devops)
- Update the `MuDevOpsWrapper` used in most repos to provide a simple `rust_build` option to opt into Rust builds
- Verify the files that will be copied to Project Mu repos for Rust support build and test in mu_plus

### Alternatives Considered

_No response_

### Urgency

Medium

### Are you going to implement the feature request?

I will implement the feature

### Do you need maintainer feedback?

No maintainer feedback needed

### Anything else?

_No response_

### PR DESCRIPTION
**Pull request by makubacki**
Closes #294 

## Description

Updates the repo to add some files needed to support Rust builds and a `RustAdvancedLoggerDxe` library crate.

Because this repo was used to test the Rust build and is adding a new crate in that process, the infrastructure files are added directly in this PR. In the future, these will be synced from `mu_devops`. The file content aligns with the current templates in `mu_devops` so there should be no change in the sync process in this repo. Future repos can simply sync the files with the bot accounts and then add the Rust code in a follow up PR.

The ability to build with Rust is temporarily disabled in the last commit of the PR branch because changes in `mu_devops` need to be completed before the pipeline build in this repo will succeed.

### Commit/Change Overview

---

#### Add initial Rust infrastructure files

Adds workspace level files to build Rust in the repo.

---

#### .azurepipelines: Add Rust build support

Makes the changes needed to build Rust in pipelines on Windows
and Linux.

---

#### AdvLoggerPkg: Add RustAdvancedLoggerDxe

Adds a library crate that serves as a Rust wrapper for access to
Advanced Logger protocol.

---

- [x] Impacts functionality?
  - **Functionality** - Does the change ultimately impact how firmware functions?
  - Examples: Add a new library, publish a new PPI, update an algorithm, ...
- [ ] Impacts security?
  - **Security** - Does the change have a direct security impact on an application,
    flow, or firmware?
  - Examples: Crypto algorithm change, buffer overflow fix, parameter
    validation improvement, ...
- [ ] Breaking change?
  - **Breaking change** - Will anyone consuming this change experience a break
    in build or boot behavior?
  - Examples: Add a new library class, move a module to a different repo, call
    a function in a new library class in a pre-existing module, ...
- [x] Includes tests?
  - **Tests** - Does the change include any explicit test code?
  - Examples: Unit tests, integration tests, robot tests, ...
- [ ] Includes documentation?
  - **Documentation** - Does the change contain explicit documentation additions
    outside direct code modifications (and comments)?
  - Examples: Update readme file, add feature readme file, link to documentation
    on an a separate Web page, ...

## How This Was Tested

- Built container image from mu_devops locally with docker
- Ran container image locally to perform a Rust build
- Ran all changes in this PR in mu_plus pipelines to verify build results
  - [Linux build with Rust](https://dev.azure.com/projectmu/mu/_build/results?buildId=52786&view=results)
  - [Linux build without Rust](https://dev.azure.com/projectmu/mu/_build/results?buildId=52788&view=results)
  - [Windows build with Rust](https://dev.azure.com/projectmu/mu/_build/results?buildId=52747&view=results)

## Integration Instructions

No changes are needed for mu_plus integrators.

Review [general Rust documentation](https://github.com/microsoft/mu_basecore/blob/HEAD/Docs/rust_build.md) in mu_basecore for more info.

### CODE REVIEW

**Review by joschock — APPROVED** [review]

File: AdvLoggerPkg/AdvLoggerPkg.ci.yaml [hunk 1]
```diff
@@ -77,5 +77,12 @@
...
         ]
+    },
+
+    "RustHostUnitTestPlugin": {
+        "Coverage": .75,
+        "CoverageOverrides": {
+            "RustAdvancedLoggerDxe": 0.0
+        }
     }
 }
```
Comment (makubacki) [inline AdvLoggerPkg/AdvLoggerPkg.ci.yaml]: I'll add a newline to this file. Want to wait for the current PR builds to finish before pushing again.

File: Cargo.lock
Comment (Javagedes) [inline Cargo.lock:1]: This file needs removed and we should add it to the gitignore

**Review by Javagedes — APPROVED** [review]

File: AdvLoggerPkg/AdvLoggerPkg.ci.yaml (hunk 1 above)
Comment (makubacki) [reply AdvLoggerPkg/AdvLoggerPkg.ci.yaml]: Done

File: Cargo.lock
Comment (makubacki) [reply Cargo.lock:1]: I think that makes sense for this repo. Pushed the update.

File: .azurepipelines/MuDevOpsWrapper.yml [hunk 2]
```diff
@@ -85,5 +93,18 @@ jobs:
     vm_image: $(vm_image)
     container_build: ${{ parameters.container_build }}
 
-- ${{ parameters.extra_jobs }}
+- ${{ if eq(parameters.rust_build, true) }}:
+  - job: CargoCmds
+    displayName: Workspace Cargo Commands
 
+    container:
+      image: ghcr.io/microsoft/mu_devops/ubuntu-22-build:9ab29bc
```
Comment (apop5) [inline .azurepipelines/MuDevOpsWrapper.yml:101]: between this and line 78, we will have two locations to update if we need to change the container image, would it make sense to localize it into a parameter?

**Review by apop5 — APPROVED** [review]

File: .azurepipelines/MuDevOpsWrapper.yml (hunk 2 above)
Comment (makubacki) [reply .azurepipelines/MuDevOpsWrapper.yml:101]: They're both updated with from the single `linux_build_container` value in [mu_devops/.sync/Version.njk](https://github.com/microsoft/mu_devops/blob/main/.sync/Version.njk) in the file sync so there's no manual maintenance needed directly in the file in this repo.

**Pull request merged**